tokio/src/util/idle_notified_set.rs

```diff
@@ -0,0 +1,463 @@
+    pub(crate) fn new() -> Self {
+        let lists = Lists {
+            inner: Mutex::new(ListsInner {
+                notified: LinkedList::new(),
+                idle: LinkedList::new(),
+                waker: None,
+            }),
+        };
+
+        IdleNotifiedSet {
+            lists: Arc::new(lists),
+            length: 0,
+        }
+    }
+
+    pub(crate) fn len(&self) -> usize {
+        self.length
+    }
+
+    pub(crate) fn is_empty(&self) -> bool {
+        self.length == 0
+    }
+
+    /// Insert the given value into the `idle` list.
+    pub(crate) fn insert_idle(&mut self, value: T) -> EntryInOneOfTheLists<'_, T> {
+        self.length += 1;
+
+        let entry = Arc::new(ListEntry {
+            parent: self.lists.clone(),
+            value: UnsafeCell::new(ManuallyDrop::new(value)),
+            my_list: UnsafeCell::new(List::Idle),
+            pointers: linked_list::Pointers::new(),
+            _pin: PhantomPinned,
+        });
+
+        {
+            let mut lock = self.lists.inner.lock();
+            lock.idle.push_front(entry.clone());
+        }
+
+        // Safety: We just put the entry in the idle list, so it is in one of the lists.
+        EntryInOneOfTheLists { entry, set: self }
+    }
+
+    /// Pop an entry from the notified list to poll it. The entry is moved to
+    /// the idle list atomically.
+    pub(crate) fn pop_notified(&mut self, waker: &Waker) -> Option<EntryInOneOfTheLists<'_, T>> {
+        // We don't decrement the length because this call moves the entry to
+        // the idle list rather than removing it.
+        if self.length == 0 {
+            // Fast path.
+            return None;
+        }
+
+        let mut lock = self.lists.inner.lock();
+
+        let should_update_waker = match lock.waker.as_mut() {
+            Some(cur_waker) => !waker.will_wake(cur_waker),
+            None => true,
+        };
+        if should_update_waker {
+            lock.waker = Some(waker.clone());
+        }
+
+        // Pop the entry.
+        match lock.notified.pop_back() {
+            Some(entry) => {
+                // Safety: We are holding the lock.
+                lock.idle.push_front(entry.clone());
+                entry.my_list.with_mut(|ptr| unsafe {
+                    *ptr = List::Idle;
+                });
+                drop(lock);
+
+                // Safety: We just put the entry in the idle list, so it is in one of the lists.
+                Some(EntryInOneOfTheLists { entry, set: self })
+            }
+            None => None,
+        }
+    }
+
+    /// Call a function on every element in this list.
+    pub(crate) fn for_each<F: FnMut(&mut T)>(&mut self, mut func: F) {
+        fn get_ptrs<T>(list: &mut LinkedList<T>, ptrs: &mut Vec<*mut T>) {
+            let mut node = list.last();
+
+            while let Some(entry) = node {
+                ptrs.push(entry.value.with_mut(|ptr| {
+                    let ptr: *mut ManuallyDrop<T> = ptr;
```

Comment:
what is this `let` doing? AFAICT, the `with_mut` closure could be simplified to just:
```rust
ptrs.push(entry.value.with_mut(|ptr| {
    ptr.cast::<T>()
}));
```
so if there's a reason it *can't* be, i'd like to know why?

Reply:
I was mostly just putting in a bunch of type hints to convince myself the pointers had the types I expected.

Reply:
ah, that's fine, then!